### DISCUSSION TITLE
File async APIs

### DISCUSSION THREAD

**Discussion by TahirAhmadov** [dotnet/runtime#63153] [Q&A]
I'm sure this question was asked before, but I couldn't find anything. Is there a plan to make all file APIs available in async mode? I'm talking about:

- Opening a file (creation of FileStream)
- FileInfo Refresh(), Delete()
- DirectoryInfo Create()
- etc.

I read somewhere that the problem is unavailability of such APIs in Windows and potentially other OSes. Is there an update from the Windows team on making this available? Can we get "fake" ones for now, to replace with "real" calls later?

**Comment by GeraudFabien** [comment]
https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/concepts/async/using-async-for-file-access

**Reply by TahirAhmadov** [reply]
That page talks about some APIs which are async now, but my question was about all the other ones.

**Comment by huoyaoyuan** [comment]
> I read somewhere that the problem is unavailability of such APIs in Windows and potentially other OSes.

This should be the exact reason.

There was proposal for asynchronous **filesystem**, instead of file. However, there lacks platform support. As far I can search, the "mobile" OSes (Android, iOS, WinRT) provide some sort of async filesystem (at least `OpenAsync`). However, I can't find any support on desktop OSes (Win32/Linux/Mac).

**Reply by huoyaoyuan** [reply]
> Is there an update from the Windows team on making this available?

I want to ask this too. @dotnet/area-system-io for information

**Comment by adamsitnik** [comment]
Unfortunately the OSes that we support don't provide a way to asynchronously open a file, not to mention other operations.

**Comment by matherm-aboehm** [comment]
Hi @TahirAhmadov
I'm curious about what exactly did you mean by your 1st point

> Opening a file (creation of FileStream)

Did you mean that operation of opening a file itself should be asynchronous?
Or did you mean to just add simplified versions of most other APIs that currently use internally the `FileStream` constructor without the `asyncMode` parameter? E.g.: `FileInfo.OpenRead()` -> `FileInfo.OpenAsyncRead()` or something like that.
Specifying this parameter in other APIs would still enable async operations on the opened file, because at least Windows does have APIs for that:
https://learn.microsoft.com/en-us/windows/win32/sync/synchronization-and-overlapped-input-and-output

**Reply by TahirAhmadov** [reply]
Hello. I meant the operation of opening a file itself. It's useful in situations when the file is on a network share.

**Reply by matherm-aboehm** [reply]
Yea, sure that would also be nice to have.

Do you know if my other idea was already mentioned somewhere?
I would also like to see easier APIs to create a `FileStream` with this async mode enabled.
